Add a new chapter "Translations" (#175)
infinum/wordpress-handbook: Translations/Localization.md

> > Localization describes the subsequent process of translating an internationalized theme/plugin. Localization is often abbreviated as l10n (because there are 10 letters between the l and the n.)

Reviewer: ```suggestion
> Localisation describes the subsequent process of translating an internationalised theme/plugin. It's often abbreviated as _l10n_ (because there are 10 letters between the _l_ and the _n_.)
```

> This is a handy up-to-date how-to for generating WordPress Theme/Plugin translation files with WP-CLI. The following workflow is covered: creating a POT (Portable Object Template) file, copying it to a PO (Portable Object) file where the `msgstr` sections are translated or updating pre-existing PO file and finally building the MO (Machine Object) file which is used by `gettext` functions to output translations to the user.

Reviewer: ```suggestion
This is a handy up-to-date how-to for generating WordPress theme/plugin translation files with the WP CLI.

 The following workflow is covered: creating a POT (Portable Object Template) file, copying it to a PO (Portable Object) file where the `msgstr` sections are translated or updating pre-existing PO file and finally building the MO (Machine Object) file which is used by `gettext` functions to output translations to the user.
```

Author: Actually, I need to add the creation of JED files here as well :)

> ### Available commands:

Reviewer: ```suggestion
### Available commands
```

- [wp i18n make-pot](https://github.com/wp-cli/i18n-command#wp-i18n-make-pot)
- [wp i18n make-json](https://github.com/wp-cli/i18n-command#wp-i18n-make-json)
- [wp i18n make-mo](https://github.com/wp-cli/i18n-command#wp-i18n-make-mo)
- [wp i18n update-po](https://github.com/wp-cli/i18n-command#wp-i18n-update-po) - ^2.4.0 WP CLI i18n-command [package](https://github.com/wp-cli/i18n-command) is required.

> ### Useful Notes

Reviewer: ```suggestion
### Useful notes
```

> - The	following files and folders are always excluded: `node_modules, .git, .svn, .CVS, .hg, vendor, Gruntfile.js, webpack.config.js, *.min.js`.

Reviewer: ```suggestion
- The following files and folders are always excluded: `node_modules, .git, .svn, .CVS, .hg, vendor, Gruntfile.js, webpack.config.js, *.min.js`.
```

- By default, the "Text Domain" header of the plugin or theme is used. If the header is not specified you'll need to add an argument to the CLI command with text-domain it should search for: `[--domain=<domain>]`.
- when naming translation files use the [WP Locales](https://make.wordpress.org/polyglots/teams/) naming convention. Example for creating files for French language translation: fr_FR.pot, fr_FR.po, fr_FR.mo.
- `[--debug]` global WP CLI option, outputs info during runtime.
- Xdebug PHP extension with `develop` mode feature turned on can cause issues when running WP CLI I18n commands on the entire theme or plugin folder. Easiest way to get around this is to turn it off for the current session by prepending `XDEBUG_MODE=debug` to the command: `XDEBUG_MODE=debug wp i18n make-po <source> [<destination>]`.
- The POT and PO files can easily be interchangeably renamed to change file types without issues.

The Eightshift theme boilerplate folder structure for language files is used for examples: `src/I18n/languages/`.

> ## WP CLI Commands

Reviewer: ```suggestion
## WP CLI commands
```

> ### Create .pot file

Reviewer: ```suggestion
### Create a `.pot` file
```

_Command_: `wp i18n make-po <source> [<destination>]`

_Example:_
```
wp i18n make-pot . src/I18n/languages/fr_FR.pot --exclude=public,assets
```

> This will scan through the entire current directory and output a fr_FR.pot file in the `src/I18n/languages` folder.

Reviewer: ```suggestion
This will scan through the entire current directory and output a `fr_FR.pot` file in the `src/I18n/languages` folder.
```

> ### Create .po file

Reviewer: ```suggestion
### Create a `.po` file
```

> Copy the .pot file in the same directory and change the extension to .po.

Reviewer: ```suggestion
Copy the `.pot` file in the same directory and change its file extension to `.po`.
```

_Example:_
```
cp src/I18n/languages/fr_FR.pot src/I18n/languages/fr_FR.po
```

> Translate `msgstr` sections.

Reviewer: ```suggestion
Translate the `msgstr` sections.
```

> ### Update .po file(s)

Reviewer: ```suggestion
### Update `.po` file(s)
```

_You may need to manually update wp i18n package (^2.4.0) for this to work: `wp package install git@github.com:wp-cli/i18n-command.git`._

If there is a pre-existing `.po` file you can update it with new entries while keeping the existing translations.

_Command_: `wp i18n update-po <source> [<destination>]`

_Example - Single:_
```
wp i18n update-po src/I18n/languages/fr_FR.pot src/I18n/languages/fr_FR.po
```

_Example - Batch:_
```
wp i18n update-po src/I18n/languages/*.pot src/I18n/languages/
```

> ### Create .json file

Reviewer: ```suggestion
### Create the JSON file
```

> If you need to make translations for javaScript files it can also be done with the CLI. Keep in mind that you shouldn't use the [--skip-js] flag when creating POT files in the first step because `.js` files need to be scanned for translations.

Reviewer: ```suggestion
If you need to make translations for JavaScript files it can also be done with the CLI.

⚠️  Keep in mind that you shouldn't use the `--skip-js` flag when creating POT files in the first step because `.js` files need to be scanned for translations.
```

Author: I'll add the inline code markup but keep the squared brackets since other flags are written like that under the "Useful notes" section.

> This command creates multiple files, one file per source `.js` file. Presumably you want to localize only block editor scripts so we will use only the one file which is generated on build from the `public/applicationBlocksEditor.js` because it contains all necessary translations per locale.

Reviewer: ```suggestion
This command creates multiple files, one file per source `.js` file. Presumably you want to localise only block editor scripts, so we will use only the one file which is generated on build from the `public/applicationBlocksEditor.js` because it contains all necessary translations per locale.
```

> It is safe to use [--purge] flag to strip `.js` related translations from the PO file as they are not useful for MO binary file.

Reviewer: ```suggestion
It is safe to use `--purge` flag to strip `.js` related translations from the PO file as they are not useful for MO binary file.
```

_Command_: `wp i18n make-json <source> [<destination>]`

_Example - Single:_
```
wp i18n make-json src/I18n/languages/fr_FR.po
```

_Example - Batch:_
```bash
wp i18n make-json src/I18n/languages/
```

> If you're using the Eightshift Libs `I18n` class there is an action hook which registers the required `.json` file via the `wp_set_script_translations` function. It presumes the file will be located in the same folder as other localization files (src/I18n/languages) and named in the following format: `{textdomain}-{locale}-block-editor-scripts.json`. So in our example that would be `{textdomain}-fr_FR-{textdomain}-block-editor-scripts.json`.

Reviewer: ```suggestion
If you're using Eightshift Libs `I18n` class there is an action hook which registers the required `.json` file via the `wp_set_script_translations` function. It presumes the file will be located in the same folder as other localisation files (src/I18n/languages) and named in the following format: `{textdomain}-{locale}-block-editor-scripts.json`. In our example that would be `{textdomain}-fr_FR-{textdomain}-block-editor-scripts.json`.
```

> Find the `.json` which contains all the translations, rename it accordingly and remove other locale related files created by the `make-json` command. This file can be easily found as it contains "source" property with a path to the main JS file of interest which in our case is `applicationBlocksEditor.js`.

Reviewer: ```suggestion
Find the `.json` which contains all the translations, rename it accordingly and remove other locale related files created by the `make-json` command. This file can be easily found as it contains the `source` property with a path to the main JS file of interest which in our case is `applicationBlocksEditor.js`.
```

You can test the result by changing the locale in the backend.

> ### Create .mo file(s)

Reviewer: ```suggestion
### Create `.mo` file(s)
```

> Last thing to do is to create the binary .mo file which is used by WordPress to output translations for the theme / plugin.

Reviewer: ```suggestion
Last thing to do is to create the binary (`.mo`) file which is used by WordPress to output translations for the theme / plugin.
```

> > Localization describes the subsequent process of translating an internationalized plugin. Localization is often abbreviated as l10n (because there are 10 letters between the l and the n.)

Reviewer 2: ```suggestion
> Localization describes the subsequent process of translating an internationalized theme or plugin. Localization is often abbreviated as l10n (because there are 10 letters between the l and the n.)
```

Author: This is a quote from [WordPress handbook](https://developer.wordpress.org/themes/functionality/localization/). Should I change it anyway?

Reviewer 2: Yes for the themes handbook. For plugins is [this](https://developer.wordpress.org/plugins/internationalization/localization/) 😂.

So adding themes/plugins is ok 🙂

### Useful Notes
- The	following files and folders are always excluded: `node_modules, .git, .svn, .CVS, .hg, vendor, Gruntfile.js, webpack.config.js, *.min.js`.
- By default, the "Text Domain" header of the plugin or theme is used. If the header is not specified you'll need to add an argument to the CLI command with text-domain it should search for: `[--domain=<domain>]`.
- when naming translation files use the [WP Locales](https://make.wordpress.org/polyglots/teams/) naming convention. Example for creating files for French language translation: fr_FR.pot, fr_FR.po, fr_FR.mo.
- `[--debug]` global WP CLI option, outputs info during runtime.
- Xdebug PHP extension with `develop` mode feature turned on can cause issues when running WP CLI I18n commands on the entire theme or plugin folder. Easiest way to get around this is to turn it off for the current session by prepending `XDEBUG_MODE=debug` to the command: `XDEBUG_MODE=debug wp i18n make-po <source> [<destination>]`.
- The POT and PO files can easily be interchangeably renamed to change file types without issues.

> 
> 
> Most of the WP CLI i18n-command defaults are fine but you might want to add these flags:

Reviewer 2: ```suggestion

Most of the WP CLI i18n-command defaults are fine but you may want to add these flags:
```

Even though may/might have similar meanings, and might describes the lesser certainty, the 'may' reads better in this case (for me at least).

Author: Now that you brought that up I'm not sure myself either but according to [this](https://www.masterclass.com/articles/may-vs-might-explained) and some other resources "might" is more suggestive and "may" is more permissive. 🤷🏼 Willing to go either way 😄

Reviewer 2: @infinum/wordpress-team what reads better to you might or may? may: 👍🏼 might: 🚀

Author: Having added the part about creating JED files to localize JS files, this note is now obsolete. We wouldn't want to exclude JS files nor the `public/` folder anymore because the main JED file is created from the `public/applicationBlocksEditor.js` file.

Reviewer 2: Ok, just to confirm, you were right. Might is more common, may is more polite.

- `[--skip-js]` to exclude .js files altogether or,
- `--exclude=public,assets` to exclude these folders.

The Eightshift theme boilerplate folder structure for language files is used for examples: `src/I18n/languages/`.

## WP CLI Commands

### Create .pot file

> >

Reviewer 2: Is this a stray quote mark?

> _Command_:\

Reviewer 2: ```suggestion
_Command_:
```

`wp i18n make-po <source> [<destination>]`

_Example:_
```
wp i18n make-pot . src/I18n/languages/fr_FR.pot --exclude=public,assets
```

> This will scan through entire current directory and output a fr_FR.pot file in the `src/I18n/languages` folder.

Reviewer 2: ```suggestion
This will scan through the entire current directory and output a fr_FR.pot file in the `src/I18n/languages` folder.
```

### Create .po file

Copy the .pot file in the same directory and change the extension to .po.

_Example:_
```
cp src/I18n/languages/fr_FR.pot src/I18n/languages/fr_FR.po
```

Translate `msgstr` sections.

### Update .po file(s)

_You may need to manually update wp i18n package (^2.4.0) for this to work: `wp package install git@github.com:wp-cli/i18n-command.git`._

If there is a pre-existing `.po` file you can update it with new entries while keeping the existing translations.

> _Command_:\

Reviewer 2: ```suggestion
_Command_:
```

`wp i18n update-po <source> [<destination>]`

_Example - Single:_
```
wp i18n update-po src/I18n/languages/fr_FR.pot src/I18n/languages/fr_FR.po
```

_Example - Batch:_
```
wp i18n update-po src/I18n/languages/*.pot src/I18n/languages/
```

### Create .mo file(s)

Last thing to do is to create the binary .mo file which is used by WordPress to output translations for the theme / plugin.

> _Command_:\ `wp i18n make-mo <source> [<destination>]`

Reviewer 2: ```suggestion
_Command_: `wp i18n make-mo <source> [<destination>]`
```
